test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,280 @@
+        input_args = ',\n'.join([arg_name for arg_type, arg_name in cpp_forward_arg_declarations])
+        for template in [TORCH_NN_MODULE_TEST_INIT, TORCH_NN_MODULE_TEST_FORWARD, TORCH_NN_MODULE_TEST_BACKWARD]:
+            cpp_source += template.substitute(
+                module_variant_name=test_params['module_variant_name'],
+                module_qualified_name=cpp_namespace + test_params['module_name'],
+                cpp_constructor_args=test_params['cpp_constructor_args'],
+                input_arg_declarations=input_arg_declarations,
+                input_args=input_args)
+        for method in torch_nn_test_methods:
+            functions.append(test_params['module_variant_name'] + '_test_' + method)
+
+        # Just-in-time compile the C++ test code
+        cpp_module = torch.utils.cpp_extension.load_inline(
+            name=test_params['module_variant_name'],
+            cpp_sources=cpp_source,
+            functions=functions,
+            verbose=False,
+        )
+
+        def test_method(method_name, test_params):
+            device = test_params['device']
+            input_size = test_params.get('input_size', None)
+            input_fn = test_params.get('input_fn', None)
+            python_module_class = test_params['python_module_class']
+            python_constructor_args = test_params['python_constructor_args']
+            module_variant_name = test_params['module_variant_name']
+            cpp_test_name = module_variant_name + '_test_' + method_name
+            if input_size:
+                example_inputs = (torch.randn(input_size), )
+            elif input_fn:
+                example_inputs = input_fn()
+            else:
+                raise RuntimeError('Missing `input_size` or `input_fn` for {}'.format(module_variant_name))
+            example_inputs = tuple([x.to(device) for x in example_inputs])
+            if method_name == 'init':
+                args = test_init(device, python_module_class, python_constructor_args)
+            elif method_name == 'forward':
+                args = test_forward(device, python_module_class, python_constructor_args, example_inputs)
+            elif method_name == 'backward':
+                args = test_backward(device, python_module_class, python_constructor_args, example_inputs)
+            # We use JIT tracing to transfer Python module state to C++
+            modules = args[0]
+            module_file_names = []
+            for module in modules:
+                traced_script_module = torch.jit.trace(module, example_inputs)
+                module_file = tempfile.NamedTemporaryFile(delete=False)
+                traced_script_module.save(module_file.name)
+                module_file.close()
+                module_file_names.append(module_file.name)
+            
+            cpp_args = []
+            for module_file_name in module_file_names:
+                cpp_args.append(module_file_name)
+            for arg in args[1:]:
```

Comment:
How dynamic `args` is makes me a little uncomfortable, but... I guess it is okay

Reply:
Yeah it's a bit unfortunate because the C++ test function has to take an unpacked list of input arguments..